### PR TITLE
Enable the RUF013 rule in ruff

### PR DESCRIPTION
**Pull request by jkseppan**
As suggested by @Skylion007 in https://github.com/pytorch/pytorch/pull/105791#issuecomment-1646644405

This rule enforces `Optional[...]` as the type for parameters having a default value of None.

Apply all RUF013 patches from ruff. Apply all other ruff patches in the affected files - otherwise lintrunner fails the build.

Fixes parts of #105230

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @Xia-Weiwen @ngimel @anijain2305

### CODE REVIEW

**Comment by Skylion007** [comment]
That was fast @malfet

File: pyproject.toml [hunk 1]
```diff
@@ -72,6 +72,7 @@ select = [
     "PERF",
     "PLE",
     "TRY302",
```
Comment (Skylion007) [inline pyproject.toml:74]: Please keep in alphabetical order (above TRY)

**Review by malfet — APPROVED** [review]

File: pyproject.toml (hunk 1 above)
Comment (jkseppan) [reply pyproject.toml]: Sure. I also moved  UP assuming the whole block after the comment should be alphabetical.

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
You have to apply RUFF013 to all files using `lintrunner --all-files`

**Comment by jkseppan** [comment]
> You have to apply RUFF013 to all files using `lintrunner --all-files`

Strangely, this didn't flag these lines as errors for me locally, with ruff 0.0.277 installed from pip. Installing 0.0.278 from conda-forge shows errors that seem to be related to importing `Optional` from `torch._jit_internal` instead of `typing` (and an unrelated noqa spelling mistake). Let's see if fixing those helps with the CI lintrunner.

**Comment by Skylion007** [comment]
Rebase the PR, I just updated Ruff.

**Comment by Skylion007** [comment]
You also need to change some the backwards compatible tests I believe

**Comment by jkseppan** [comment]
Here's one of the errors:

```
FAILED [0.0122s] test_fx.py::TestFXAPIBackwardCompatibility::test_function_back_compat - AssertionError: "torc[8771 chars]_fn: Callable[[torch.fx.node.Node], Proxy] = N[515 chars]h]\n" != "torc[8771 chars]_fn: Optional[Callable[[torch.fx.node.Node], P[525 chars]h]\n"
[...]
- torch.fx.proxy.TracerBase.create_proxy(self, kind: str, target: torch.fx.node.Target, args: Tuple[Any, ...], kwargs: Dict[str, Any], name: Optional[str] = None, type_expr: Optional[Any] = None, proxy_factory_fn: Callable[[torch.fx.node.Node], Proxy] = None)
+ torch.fx.proxy.TracerBase.create_proxy(self, kind: str, target: torch.fx.node.Target, args: Tuple[Any, ...], kwargs: Dict[str, Any], name: Optional[str] = None, type_expr: Optional[Any] = None, proxy_factory_fn: Optional[Callable[[torch.fx.node.Node], Proxy]] = None)
?                                                                                                                                                                                                                     +++++++++                                     +
[...]
****** ERROR ******
An FX function that has been marked as backwards-compatible has experienced a signature change. See the above exception context for more information. If this change was unintended, please revert it. If it was intended, check with the FX team to ensure that the proper deprecation protocols have been followed and subsequently --accept the change.

To execute this test, run the following from the base repo dir:
     python test/test_fx.py -k test_function_back_compat
```

The difference is in `create_proxy`, where the argument `proxy_factory_fn` has changed type from `Callable[...]` to `Optional[Callable[...]]` because it has a default value of `None`. I think the non-Optional type is wrong, but the message is asking me to check with the FX team. How would I go about doing that?

**Comment by jkseppan** [comment]
Many of the other errors are due to the type of `adagrad`'s argument `has_sparse_grad` changing from `bool` to `Optional[bool]`. It seems to me that the default value of `None` should instead be `False`, and then the type would not need to change. But I may be overlooking something here - in the `foreach` argument the `None` value seems to mean something distinct from `False`.

**Comment by jkseppan** [comment]
I rebased to fix the conflict, changed a backward-compatibility test to expect Optional in the signature and added two noqa comments because of a Ruff false positive. The backward-compatibility test told me to "check with the FX team" - how can I contact the FX team?

**Comment by Skylion007** [comment]
@janeyx99 Any idea what to do the about the optimizer signature failing tests?

**Comment by janeyx99** [comment]
The optimizer bool optional bool errors are due to torchscript not able to deduce the value of has_sparse_grad. The way to go about this is to explicitly set the variable to False if it is found to be None, similar to what we do for foreach where we have:

```
if foreach is None:
    foreach = False
```

**Comment by jkseppan** [comment]
I rebased on top of current main, changed the adagrad has_sparse_grad handling so it can be an Optional[bool] and added seemingly unrelated RUF017 fixes, because Ruff was failing on those.

**Pull request closed without merging**